File: Airwallex/AirwallexCore/Sources/Model/AWXAddress.m
//
//  AWXAddress.m
//  Airwallex
//
//  Created by Victor Zhu on 2020/1/14.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "AWXAddress.h"

@implementation AWXAddress

- (NSDictionary *)encodeToJSON {
    return @{
        @"country_code": self.countryCode,
        @"city": self.city ?: @"",
        @"street": self.street ?: @"",
        @"state": self.state ?: @"",
        @"postcode": self.postcode ?: @""
    };
}

+ (id)decodeFromJSON:(NSDictionary *)json {
    AWXAddress *address = [AWXAddress new];
    address.countryCode = json[@"country_code"];
    address.city = json[@"city"];
    address.street = json[@"street"];
    address.state = json[@"state"];
    address.postcode = json[@"postcode"];
    return address;
}

- (id)copyWithZone:(nullable NSZone *)zone {
    AWXAddress *copy = [[AWXAddress allocWithZone:zone] init];
    copy.countryCode = [self.countryCode copyWithZone:zone];
    copy.city = [self.city copyWithZone:zone];
    copy.street = [self.street copyWithZone:zone];
    copy.state = [self.state copyWithZone:zone];
    copy.postcode = [self.postcode copyWithZone:zone];
    return copy;
}

@end

File: Airwallex/Core/Sources/Internal/Models/AWXAddress.swift
//
//  AWXAddress.swift
//  Core
//
//  Created by Tony He (CTR) on 2024/7/19.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

/// `AWXAddress` includes the information of an address.
@objcMembers
@objc
public class AWXAddress: NSObject, Codable {
    /**
     Country code of the address. Use the two-character ISO Standard Country Codes.
     */
    public let countryCode: String?

    /**
     City of the address.
     */
    public let city: String?

    /**
     Street of the address.
     */
    public let street: String?

    /**
     State or province of the address, optional.
     */
    public let state: String?

    /**
     Postcode of the address, optional.
     */
    public let postcode: String?

    enum CodingKeys: String, CodingKey {
        case countryCode = "country_code"
        case city
        case street
        case state
        case postcode
    }

    public init(countryCode: String?, city: String?, street: String?, state: String?, postcode: String?) {
        self.countryCode = countryCode
        self.city = city
        self.street = street
        self.state = state
        self.postcode = postcode
    }
}
